File: 543. Diameter of Binary Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
  let result = 0;
  function dfs(current){
    if (current == undefined){
      return 0;
    }
    let left = dfs(current.left);
    let right = dfs(current.right);
    
    if (left + right > result){
      result = left + right;
    }

    let maximum = left;
    if (right > maximum){
      maximum = right;
    }
    return maximum + 1;
  }
  
  dfs(root);
  return result;
};

File: 543. Diameter of Binary Tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  int result = 0;
  int dfs(TreeNode* current){
    if (current == nullptr){
      return 0;
    }
    int left = dfs(current->left);
    int right = dfs(current->right);
    
    if (left + right > result){
      result = left + right;
    }

    int maximum = left;
    if (right > maximum){
      maximum = right;
    }
    return maximum + 1;
  }
public:
  int diameterOfBinaryTree(TreeNode* root) {
    dfs(root);
    return result;
  }
};